### PR TITLE
wireguard-tools: add ippeer option

### PR DESCRIPTION
**Pull request by champtar**
This allow to set IPv4 peer address for point to point tunnel
This simplify a lot bird ospf usage / configuration
```
~# cat /etc/config/network
config interface 'test'
	option proto 'wireguard'
	option private_key '<key>'
	list addresses '1.2.3.4'
	option ippeer '5.6.7.8'
	option nohostroute '1'

~# ip a show test
9: test: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN qlen 1000
    link/[65534]
    inet 1.2.3.4 peer 5.6.7.8/32 brd 255.255.255.255 scope global test
       valid_lft forever preferred_lft forever
```
------------------------------------------------

`ippeer` is not really a good fit with `addresses` but I don't have a better idea for now.

### CODE REVIEW

**Comment by champtar** [comment]
@zx2c4

File: package/network/utils/wireguard-tools/files/wireguard.sh [hunk 1]
```diff
@@ -108,6 +108,7 @@ proto_wireguard_setup() {
 	config_get private_key "${config}" "private_key"
 	config_get listen_port "${config}" "listen_port"
 	config_get addresses "${config}" "addresses"
+	config_get ippeer "${config}" "ippeer"
```
Comment (dedeckeh) [inline package/network/utils/wireguard-tools/files/wireguard.sh:111]: For the sake of uniformity I propose to change this into peeraddr; similar as in the packages ipip  (https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/network/config/ipip/files/ipip.sh;h=15b1c978e31b1809021ce15aef4aa656b23886b7;hb=refs/heads/master#l85) and gre (https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/network/config/gre/files/gre.sh;h=eb3df5b48c874c7e14b32bd53bb27a8ff0854684;hb=refs/heads/master#l269)

Comment (champtar) [reply package/network/utils/wireguard-tools/files/wireguard.sh:111]: `peerip` could be named `endpoint_ip`, and is not at all similar to `ippeer`
`ippeer` is a generic option that could be available for any interface type
```
ip link add name ipip-test type ipip local 9.10.11.12 remote 13.14.15.16
ip addr add dev ipip-test local 1.2.3.4 peer 5.6.7.8
ip addr show ipip-test
15: ipip-test@NONE: <POINTOPOINT,NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 9.10.11.12 peer 13.14.15.16
    inet 1.2.3.4 peer 5.6.7.8/32 scope global ipip-test
       valid_lft forever preferred_lft forever
```

Comment (champtar) [reply package/network/utils/wireguard-tools/files/wireguard.sh:111]: Really good article about tunnel interfaces: https://developers.redhat.com/blog/2019/05/17/an-introduction-to-linux-virtual-interfaces-tunnels/

**Comment by zx2c4** [comment]
I'm curious to learn how this simplifies bird and ospf. How does it differ from setting the IP address to a /32 and adding one single interface route of a /32?

**Comment by champtar** [comment]
```
# ip link add name wg-test type wireguard
# ip addr add dev wg-test local 1.2.3.4
# ip link set wg-test up

**Comment by champtar** [comment]
Extract from bird conf
```
	area 0.0.0.0 {
		stub no;
		stubnet 10.255.3.3/32;
		interface "wg*" {
			cost 40;
			transmit delay 5;
			dead count 5;
			authentication cryptographic;
			password "<thepassword>";
		};
	};
```

**Comment by zx2c4** [comment]
What I'm wondering about is:

```
# ip link add name wg-test type wireguard
# ip link set wg-test up
# ip addr add 1.2.3.4/32 dev wg-test
# ip route add 5.6.7.8/32 dev wg-test
```

**Comment by champtar** [comment]
Sorry I did a bunch of edits, you should have your response in previous messages

It doesn't change the routing, and bird/ospf still uses multicast
```
# timeout 30s tcpdump -i wg-test -nnev
dropped privs to tcpdump
tcpdump: listening on wg-test, link-type RAW (Raw IP), capture size 262144 bytes
15:52:12.149201 ip: (tos 0xc0, ttl 1, id 42808, offset 0, flags [none], proto OSPF (89), length 80)
    1.2.3.4 > 224.0.0.5: OSPFv2, Hello, length 44
	Router-ID 54.36.54.117, Backbone Area, Authentication Type: MD5 (2)
	Key-ID: 1, Auth-Length: 16, Crypto Sequence Number: 0x600d883d
	Options [External]
	  Hello Timer 10s, Dead Timer 50s, Mask 0.0.0.0, Priority 1
```
But this is how bird/ospf works with point to point /32 links, if there is no peer set it's a stub

**Comment by zx2c4** [comment]
Huh, interesting. Is this something that should be fixed in bird? Or is the `peer` address an important semantic worth supporting in OpenWRT because it has special significance beyond this particular bug?

**Comment by champtar** [comment]
I have no idea if this is some legacy behavior or important semantic. I've migrated from openvpn p2p tunnel (with /32 & peer) to wireguard last summer and was forced to use /31 and I really prefer /32, it makes the setup way cleaner/easier IMO.
I'll look a bit at bird code and send an email to bird mailing list with you in CC.

**Comment by champtar** [comment]
Email sent (http://trubka.network.cz/pipermail/bird-users/2021-January/015157.html)
might just be historical https://unix.stackexchange.com/a/198507

**Comment by champtar** [comment]
Answer: this is a bird limitation but pretty hard to fix (http://trubka.network.cz/pipermail/bird-users/2021-January/015167.html)

**Comment by champtar** [comment]
@zx2c4 can we merge this ? Between rewriting bird and a 3 lines patch, I prefer the 3 lines patch :)

**Comment by champtar** [comment]
@zx2c4 friendly ping

**Comment by zx2c4** [comment]
Isn't this a small patch for bird too? Why a "rewrite"?

**Comment by champtar** [comment]
> BIRD OSPF implementation is strongly coupled with address slots instead of real interfaces.
This makes me think one will need to change a good part of the logic of BIRD OSPF implementation, not just add one or 2 if.
With my skill level in C, my free time and no one offering to do the bird changes, please pretty please accept this feature addition.

**Comment by champtar** [comment]
@zx2c4 ?

**Comment by champtar** [comment]
@zx2c4 silence is hard to parse ;)

**Comment by champtar** [comment]
Rebased on latest changes
@zx2c4 could you ack/nack please

**Comment by zx2c4** [comment]
> With my skill level in C, my free time and no one offering to do the bird changes, please pretty please accept this feature addition.

This isn't a good rationale for adding a non standard knob that will harder to claw back later when we risk breaking compatibility. I'll feel comfortable acking this if I'm more certain that there's no way for bird to do it, or that there's some important semantic place for ippeer. But if the reason amounts to, "because shell scripting is easier than C", I don't want to start taking shortcuts like that.

Basically, either me or you or somebody else needs to find the time to really look deeply into this.

@tohojo mentioned an interest in adding various forms of WireGuard support to bird at some point. Maybe he has a better idea of the architecture involved.

**Comment by champtar** [comment]
After double checking, ippeer is not present in NetworkManager either, so improving bird make sense.
I'll let @tohojo comment and close this.

**Comment by tohojo** [comment]
I don't really know anything about the OSPF implementation in Bird apart from the reply you linked above, sorry.

As for adding Wireguard support to bird itself, my idea for that was to teach Bird about the wireguard interface type so it could both learn peers from it, and also update allowedips when it installs new routes. Which would make it possible to run Bird on a multi-peer wireguard link instead of using p2p links. That seems rather orthogonal to this, though...

**Comment by champtar** [comment]
Thanks @tohojo and @zx2c4, I'll close this for now
